src/dbnode/storage/series/buffer.go

```diff
@@ -181,287 +161,281 @@ func (b *dbBuffer) Write(
+	if err != nil {
+		return nil, err
 	}
 
-	if b.buckets[idx].needsReset(start) {
-		// Reset bucket
-		b.buckets[idx].resetTo(start)
+	streams := buckets.streams(ctx, WarmWrite)
+	// We may need to merge again here because the regular merge method does
+	// not merge buckets that have different versions.
+	numStreams := len(streams)
+	if numStreams == 1 {
+		return streams[0], nil
 	}
 
-	return mergedOutOfOrderBlocks
-}
+	sr := make([]xio.SegmentReader, 0, numStreams)
+	for _, stream := range streams {
+		sr = append(sr, stream.SegmentReader)
+	}
 
-func (b *dbBuffer) Bootstrap(bl block.DatabaseBlock) error {
-	blockStart := bl.StartTime()
-	bootstrapped := false
-	for i := range b.buckets {
-		if b.buckets[i].start.Equal(blockStart) {
-			if b.buckets[i].drained {
-				return fmt.Errorf(
-					"block at %s cannot be bootstrapped by buffer because its already drained",
-					blockStart.String(),
-				)
-			}
-			b.buckets[i].bootstrap(bl)
-			bootstrapped = true
-			break
+	bopts := b.opts.DatabaseBlockOptions()
+	encoder := bopts.EncoderPool().Get()
+	encoder.Reset(blockStart, bopts.DatabaseBlockAllocSize())
+	iter := b.opts.MultiReaderIteratorPool().Get()
+	defer func() {
+		encoder.Close()
+		iter.Close()
+	}()
+	iter.Reset(sr, blockStart, b.opts.RetentionOptions().BlockSize())
+
+	for iter.Next() {
+		dp, unit, annotation := iter.Current()
+		if err := encoder.Encode(dp, unit, annotation); err != nil {
+			return nil, err
 		}
 	}
-	if !bootstrapped {
-		return fmt.Errorf("block at %s not contained by buffer", blockStart.String())
+	if err := iter.Err(); err != nil {
+		return nil, err
 	}
-	return nil
+
+	return encoder.Stream(), nil
 }
 
-// forEachBucketAsc iterates over the buckets in time ascending order
-// to read bucket data
-func (b *dbBuffer) forEachBucketAsc(fn func(*dbBufferBucket)) {
-	for i := 0; i < bucketsLen; i++ {
-		idx := (b.pastMostBucketIdx + i) % bucketsLen
-		fn(&b.buckets[idx])
+func (b *dbBuffer) Flush(
+	ctx context.Context,
+	blockStart time.Time,
+	id ident.ID,
+	tags ident.Tags,
+	persistFn persist.DataFn,
+	version int,
+) (FlushOutcome, error) {
+	buckets, exists := b.bucketsAt(blockStart)
+	if !exists {
+		return FlushOutcomeBlockDoesNotExist, nil
 	}
-}
 
-// computedForEachBucketAsc performs a fn on the buckets in time ascending order
-// and returns the sum of the number returned by each fn
-func (b *dbBuffer) computedForEachBucketAsc(
-	op computeBucketIdxOp,
-	fn func(now time.Time, b *dbBuffer, idx int, bucketStart time.Time) int,
-) int {
-	now := b.nowFn()
-	pastMostBucketStart := now.Truncate(b.blockSize).Add(-1 * b.blockSize)
-	bucketNum := (pastMostBucketStart.UnixNano() / int64(b.blockSize)) % bucketsLen
-	result := 0
-	for i := int64(0); i < bucketsLen; i++ {
-		idx := int((bucketNum + i) % bucketsLen)
-		curr := pastMostBucketStart.Add(time.Duration(i) * b.blockSize)
-		result += fn(now, b, idx, curr)
+	// A call to flush can only be for warm writes
+	block, err := buckets.toBlock(WarmWrite)
```

Comment:
Do we need to convert it to a block just so we can flush it?

Also, what happens to bootstrapped blocks here? Seems like they'll just get lost

Reply:
If I'm right about the bootstrapped blocks being lost we'll probably need an integration test for that which would basically be the same as the two ones you deleted which were no longer applicable (for rotation)

Reply:
Good catch, will fix.

Reply:
Actually, this should contain the bootstrapped blocks as well.

When a block gets bootstrapped, the writable bucket will store that block. The `toBlock` function will then return a block encapsulating all the data within the bucket, merging bootstrapped blocks and encoders. Sometimes there's no bootstrapped block to merge with in which case we're just converting it to a block.